core/input/input.cpp

```diff
@@ -697,23 +699,32 @@ void Input::_parse_input_event_impl(const Ref<InputEvent> &p_event, bool p_is_em
 	for (const KeyValue<StringName, InputMap::Action> &E : InputMap::get_singleton()->get_action_map()) {
 		if (InputMap::get_singleton()->event_is_action(p_event, E.key)) {
 			Action &action = action_state[E.key];
-			// If not echo and action pressed state has changed
-			if (!p_event->is_echo() && is_action_pressed(E.key, false) != p_event->is_action_pressed(E.key)) {
+			bool is_pressed = false;
+
+			if (!p_event->is_echo()) {
 				if (p_event->is_action_pressed(E.key)) {
-					action.pressed = true;
-					action.pressed_physics_frame = Engine::get_singleton()->get_physics_frames();
-					action.pressed_process_frame = Engine::get_singleton()->get_process_frames();
+					action.pressed++;
+					is_pressed = true;
+					print_line("+", E.key, p_event, action.pressed);
+					if (action.pressed == 1) {
+						action.pressed_physics_frame = Engine::get_singleton()->get_physics_frames();
+						action.pressed_process_frame = Engine::get_singleton()->get_process_frames();
+					}
 				} else {
-					action.pressed = false;
-					action.released_physics_frame = Engine::get_singleton()->get_physics_frames();
-					action.released_process_frame = Engine::get_singleton()->get_process_frames();
+					if (action.pressed == 1) {
+						action.released_physics_frame = Engine::get_singleton()->get_physics_frames();
+						action.released_process_frame = Engine::get_singleton()->get_process_frames();
+					}
+					action.pressed = MAX(action.pressed - 1, 0);
+					print_line("-", E.key, p_event, action.pressed);
 				}
-				action.strength = 0.0f;
-				action.raw_strength = 0.0f;
 				action.exact = InputMap::get_singleton()->event_is_action(p_event, E.key, true);
 			}
-			action.strength = p_event->get_action_strength(E.key);
-			action.raw_strength = p_event->get_action_raw_strength(E.key);
+
+			if (is_pressed || action.pressed == 0) {
+				action.strength = p_event->get_action_strength(E.key);
+				action.raw_strength = p_event->get_action_raw_strength(E.key);
```

Comment:
This implementation changes the action.strength on every event, which might lead to alternating values, if two axis inputs are mapped to the same action and both axis are moved.

The comment https://github.com/godotengine/godot/issues/45628#issuecomment-782840254 suggests to use as strength the maximum strength of all related actions. Perhaps that would be a better solution?

Reply:
But then what happens on release? It should assign max strength of still pressed actions, but they are not tracked. This means that if you use analog and then press a key, your analog strength will be ignored until it's released.
Maybe it's fine, idk.

Reply:
> It should assign max strength of still pressed actions, but they are not tracked

Perhaps we could track it with the following, but I haven thought it through yet.

``` diff
diff --git a/core/input/input_map.h b/core/input/input_map.h
index b4d5beacb3..b59e898730 100644
--- a/core/input/input_map.h
+++ b/core/input/input_map.h
@@ -48,10 +48,17 @@ public:
         */
        static int ALL_DEVICES;
 
+       struct EventWithStrength {
+               Ref<InputEvent> event;
+               int pressed = 0;
+               float strength = 0.0f;
+               float raw_strength = 0.0f;
+       };
+
        struct Action {
                int id;
                float deadzone;
-               List<Ref<InputEvent>> inputs;
+               List<EventWithStrength> inputs;
        };
 
 private:
```

Reply:
This can be addressed later if someone opens an issue. Let's see if the current behavior is ok for now.

Reply:
Sounds good to me.